homeassistant/components/hassio/repairs.py

```diff
@@ -0,0 +1,142 @@
+    ATTR_UNHEALTHY,
+    ATTR_UNHEALTHY_REASONS,
+    ATTR_UNSUPPORTED,
+    ATTR_UNSUPPORTED_REASONS,
+    ATTR_UPDATE_KEY,
+    ATTR_WS_EVENT,
+    DOMAIN,
+    EVENT_HEALTH_CHANGED,
+    EVENT_SUPERVISOR_EVENT,
+    EVENT_SUPERVISOR_UPDATE,
+    EVENT_SUPPORTED_CHANGED,
+    INFO_URL_UNHEALTHY,
+    INFO_URL_UNSUPPORTED,
+    ISSUE_ID_UNHEALTHY,
+    ISSUE_ID_UNSUPPORTED,
+    UPDATE_KEY_SUPERVISOR,
+)
+from .handler import HassIO
+
+
+class SupervisorRepairs:
+    """Create repairs from supervisor events."""
+
+    def __init__(self, hass: HomeAssistant, client: HassIO) -> None:
+        """Initialize supervisor repairs."""
+        self._hass = hass
+        self._client = client
+        self._remove_dispatcher: Callable[[], None] | None = None
+        self._unsupported_reasons: set[str] = set()
+        self._unhealthy_reasons: set[str] = set()
+
+    @property
+    def unhealthy_reasons(self) -> set[str]:
+        """Get unhealthy reasons. Returns empty set if system is healthy."""
+        return self._unhealthy_reasons
+
+    @unhealthy_reasons.setter
+    def unhealthy_reasons(self, reasons: set[str]) -> None:
+        """Set unhealthy reasons. Create or delete repairs as necessary."""
+        if len(reasons) > len(self.unhealthy_reasons):
+            for unhealthy in reasons - self.unhealthy_reasons:
+                async_create_issue(
+                    self._hass,
+                    DOMAIN,
+                    f"{ISSUE_ID_UNHEALTHY}_{unhealthy}",
+                    is_fixable=False,
+                    learn_more_url=f"{INFO_URL_UNHEALTHY}/{unhealthy}",
+                    severity=IssueSeverity.CRITICAL,
+                    translation_key="unhealthy",
+                    translation_placeholders={"reason": unhealthy},
+                )
+        else:
+            for fixed in self.unhealthy_reasons - reasons:
+                async_delete_issue(self._hass, DOMAIN, f"{ISSUE_ID_UNHEALTHY}_{fixed}")
+
+        self._unhealthy_reasons = reasons
+
+    @property
+    def unsupported_reasons(self) -> set[str]:
+        """Get unsupported reasons. Returns empty set if system is supported."""
+        return self._unsupported_reasons
+
+    @unsupported_reasons.setter
+    def unsupported_reasons(self, reasons: set[str]) -> None:
+        """Set unsupported reasons. Create or delete repairs as necessary."""
+        if len(reasons) > len(self.unsupported_reasons):
+            for unsupported in reasons - self.unsupported_reasons:
+                async_create_issue(
+                    self._hass,
+                    DOMAIN,
+                    f"{ISSUE_ID_UNSUPPORTED}_{unsupported}",
+                    is_fixable=False,
+                    learn_more_url=f"{INFO_URL_UNSUPPORTED}/{unsupported}",
+                    severity=IssueSeverity.WARNING,
+                    translation_key="unsupported",
+                    translation_placeholders={"reason": unsupported},
+                )
+        else:
```

Comment:
I think we need to create issues for new reasons and remove issues for stale reasons regardless of the number of reasons. There may be non overlapping items that affect the number of the new or existing reasons.

I suggest adding a test for this case.

Reply:
True, the length check could cause a bug if one is added and removed at the same time. Good catch, will do.